lib/src/postgrest_builder.dart

```diff
@@ -210,30 +258,122 @@ class PostgrestBuilder<T> {
         count: 0,
       );
     } else {
-      return PostgrestResponse(
-        status: response.statusCode,
-        error: error,
-      );
+      throw error;
     }
   }
 
   /// Update Uri queryParameters with new key:value
   /// Use lists to allow multiple values for the same key
   void appendSearchParams(String key, String value) {
-    final searchParams = Map<String, dynamic>.from(url.queryParametersAll);
+    final searchParams = Map<String, dynamic>.from(_url.queryParametersAll);
     searchParams[key] = [...searchParams[key] ?? [], value];
-    url = url.replace(queryParameters: searchParams);
+    _url = _url.replace(queryParameters: searchParams);
   }
 
   /// Overrides Uri queryParameters with new key:value
   void overrideSearchParams(String key, String value) {
-    final searchParams = Map<String, dynamic>.from(url.queryParametersAll);
+    final searchParams = Map<String, dynamic>.from(_url.queryParametersAll);
     searchParams[key] = value;
-    url = url.replace(queryParameters: searchParams);
+    _url = _url.replace(queryParameters: searchParams);
+  }
+
+  @override
+  Stream<T?> asStream() {
+    final controller = StreamController<T?>.broadcast();
+
+    then((value) {
+      controller.add(value);
+    }).catchError((Object error, StackTrace stack) {
+      controller.addError(error, stack);
+    }).whenComplete(() {
+      controller.close();
+    });
+
+    return controller.stream;
+  }
+
+  @override
+  Future<T> catchError(Function onError, {bool Function(Object error)? test}) {
+    throw UnimplementedError('catchError should not be called in this future');
+  }
+
+  /// Register callbacks to be called when this future completes.
+  @override
+  Future<R> then<R>(
+    FutureOr<R> Function(T? value) onValue, {
+    Function? onError,
+  }) async {
+    if (onError != null &&
+        onError is! Function(Object, StackTrace) &&
+        onError is! Function(Object)) {
+      throw ArgumentError.value(
+        onError,
+        "onError",
+        "Error handler must accept one Object or one Object and a StackTrace"
+            " as arguments, and return a value of the returned future's type",
+      );
+    }
+
+    try {
+      final response = await _execute();
+      // ignore: null_check_on_nullable_type_parameter
+      final data = response.data;
+
+      if (_converter != null) {
+        assert(
+          !(_options?.forceResponse ?? false),
+          'converter and forceReponse can not be set at the same time',
+        );
+        return onValue(data as T);
+      } else {
+        if ((_options?.forceResponse ?? false) || response.count != null) {
+          return onValue(response as T);
+        } else {
+          return onValue(data as T);
+        }
+      }
+    } catch (error, stack) {
+      if (onError != null) {
+        if (onError is Function(Object, StackTrace)) {
+          onError(error, stack);
+        } else if (onError is Function(Object)) {
+          onError(error);
```

Comment:
I have been taking a deeper look into this, and I don't really have an solution to this problem, but when we call `onError`, it should not throw any errors, but instead just stop the function there, isn't it? 

When I try to use `onError` on a regular future that throws, onError does get called, but no error is thrown. 

With that being said, I don't know how implementing that is possible. I have been struggling with it for quite some time. 

Not sure if the tests are failing because of this, but I think the rework here is all very amazing things, and we should not let some tests that are failing for unknown reason prevent this from being released. We will slowly fix bugs as they come up.

Reply:
Fixed merge issues now